### ISSUE TITLE
Mouse controls of tmux don't work via a MOSH connect

### ISSUE THREAD

**Issue by nils-tekampe** [blinksh/blink#395]
Hi,
I love your app! brings me a real terminal to my iPad :-)
I often use MOSH to cope with limited bandwidth and frequent interrupts in our mobile networks and have tmux running on my server. tmux splits the screen in multiple panels and one can enable mouse control to use the mouse to select the currently active panel. However, this does not work via blink. Is there anything special in mouse controls? Or can I do anything to makes this work?
Thanks
Nils

**Comment by carloscabanero** [comment]
Mouse controls for Blink, whether on tmux or other apps is a must. Changing the title.

**Comment by piyush-nimbalkar** [comment]
Yes, it would be a great to have feature. Especially with tmux scrolling with mouse mode on.

**Comment by schappim** [comment]
Mouse controls would be fantastic for the Micro Text Editor https://github.com/zyedidia/micro

**Comment by nils-tekampe** [comment]
I just wanted to ask whether there are any plans on this feature. Any idea whether/when it will be implemented?

**Comment by yury** [comment]
Basic touch reports are in testing now (TestFlight). 

We are currently limited to clicks. So tmux panel selection is working. But selecting and scrolling are tricky to implement, because they conflicting with general scroll and text selection. We can’t figure out the best way to do it yet. 

If you have any suggestions, please share.

**Comment by nils-tekampe** [comment]
Great to hear that!
For scrolling I would think that a two finger swipe could help to figure out whether the user wants to scroll within the terminal or the application. But I have no idea, how I would implement the text selection part. For me it would be OK if I would be required to double-click with two fingers or so before I can actually select text within a terminal. But I'm not sure whether this is really intuitive.

**Comment by yury** [comment]
Two thinger swipe is for kb dismiss and control panel. Also two fingers tap open new session. Also two fingers drag sometimes conficts with zoom. So we should find better solution.

**Comment by nils-tekampe** [comment]
I see. Seems to be really tricky, what about a “catch mode” similar to how Parallels or VirtualBox work. As long as I’m caught in the terminal, all mouse gestures go to the terminal. If I want to control the app, I have to release the mode first (e.g. by the use of a certain gesture or button). Definitely less intuitive but could work.

**Comment by bradleyjones** [comment]
“Catch mode” could be interesting, that would allow you to use familiar gestures for scrolling (one/two fingers) without interferring with the rest of the app. Failing that even just a 3 finger scroll would be ok with me and just scroll whatever pane has focus. Personally tmux scrolling is more important than selection, I’m quite happy just making a panel fullscreen and selecting text using the normal methods.

**Comment by yury** [comment]
What about this way?

https://youtu.be/UZLjhSPsIVY

Need to find a way to do that with hardware kb..

**Comment by nils-tekampe** [comment]
Not bad :-) Can I also click in to the separate windows to set the focus for the cursor to them?

**Comment by yury** [comment]
@nils-tekampe you can do that in current version already. Just tap on the pane you need to focus on.

**Comment by nils-tekampe** [comment]
Great to hear. I stopped using blinksh after I learned that his was not working in the version that I tried at this time. Will give it a try later!

**Comment by NyaMisty** [comment]
It seems that this issue and #119 #566 #584 are all about the mouse scroll support. Which one should we use to track the progress?

**Comment by willpuckett** [comment]
Hi I have mouse clicks working with this in VIM (set mouse=a). I am wondering if a meta key for scrolling was ever decided on? Can't seem to locate anything.....

**Comment by arkku** [comment]
One mouse control that is still missing is drag, e.g., to select text in split tmux panes rather than in the whole terminal. I understand that this conflicts with text selection, but I wonder if it would be doable with some modifier key to override the selection and instead just pass the dragging to the terminal? This behaviour could even be made invertable, i.e., pass by default and only select if a modifier is held…

**Comment by aabdellah** [comment]
I think an option to disable text selection or make it only work with a modifier would be very welcome as it conflicts with the following tmux functionalities:
- Dragging with the left button on a pane border resizes the pane.
- Dragging with the left button inside a pane selects text; the selected text is copied when the mouse is released.
- Pressing the right button on a pane opens a menu with various commands. When the mouse button is released, the selected command is run with the pane as target. Each menu item also has a key shortcut shown in brackets.
- Pressing the right button on a window or on the session name on the status line opens a similar menu for the window or session.

**Comment by nohr** [comment]
i'm surprised to see this dead after three years. mouse drag is so crucial for me in vim and tmux. is there any chance there might be an update on that? currently iSH has it implemented pretty well but i'd much rather use blink for its simplicity...

**Comment by carloscabanero** [comment]
Honestly, we haven't gotten much requests for mouse drag controls. Most people just use the keyboard for that, plus it creates a mode (need to disable selection so all input will go to the remote).

That being said, we are moving to a different terminal emulator and this should be tackled then.

I will separate in a different issue so this is more specific.

**Comment by metawops** [comment]
Just adding myself here because I’d *really* like Blink to have full mouse controls. As soon as I’ll be able to drag pane split line in my tmux session with the trackpad of my Apple hardware keyboard I’m happy! 
Any news on this, @carloscabanero? 😳 Finished the “move to a different terminal emulator” already? And which issue should I track to stay informed on this topic? 🙏

**Comment by carloscabanero** [comment]
No, moving to a different terminal emulator has not been started yet and it is not an easy task. Other critical components like Keyboard, UI, gestures, etc... depend on it. 

Adding to this, the only mouse controls not working are the drag operations. These also collide with selection, and if I recall we would need to have a special mode to implement them.

This is a good issue to track. Once it becomes a priority, when there are multiple issues that can be closed at the same time by a "parent" issue, I try to keep all of them updated.

**Comment by brendandebeasi** [comment]
@carloscabanero I found that https://github.com/holzschu/a-shell seems to have decent trackpad, mouse support (including right click!!). It's OSS so maybe you can get some inspiration from it. Thanks as always for your great work!

**Comment by emaadmanzoor** [comment]
Here to report the same issue after switching from Termius to Blink. I do sometimes copy-paste text from a tmux pane using my mouse, and for some reason the selection spans multiple panes in Blink. Screenshot for reference.

<img width="2732" height="2048" alt="Image" src="https://github.com/user-attachments/assets/9b7b0f9d-e7c5-47fc-9301-1157e4b3ec45" />

**Comment by carloscabanero** [comment]
Implemented on 18.5.0! 

We implemented mouse drag trying to replicate the "mode" you usually have with the keyboard, so you can have more precise cursor positioning, for both touch and the trackpad.
  - Trackpad: Cmd+click to start drag mode, move cursor while holding Cmd, release to stop.
  - Touch: Long press to start drag mode and place the cursor, then start a drag to move cursor and release to stop. 
  - Remember you can use the keyboard while you drag.

**Comment by brendandebeasi** [comment]
@carloscabanero bravo! Any plans to support modifier Keyes with the drag for different selection modes, specifically for TUI?

**Comment by carloscabanero** [comment]
Not sure I follow, but can you give me an example of what you need? I can definitely make tweaks before we ship.

**Comment by brendandebeasi** [comment]
Essentially, TUI apps let you hold option to select text within the context of the UI.

So if you were trying to copy text, that is multiple lines, but formatted into several columns, you could hold option and drag to select the text just in that single column even though it is multiple lines.

I am AFK at the moment, but I hope that makes sense. Let me know if not, and I can upload a video.

**Comment by carloscabanero** [comment]
Yeah, that is exactly what this fix is about :)

**Comment by emaadmanzoor** [comment]
This is amazing! Looking forward to the update.

**Comment by carloscabanero** [comment]
This should be out on 18.5.0.
